### ISSUE TITLE
run watch updates app.js twice, then compiles successfully without changing app.js when files change

### ISSUE THREAD

**Issue by eveames** [JeffreyWay/laravel-mix#2902]
-   Laravel Mix Version: 6.0.10 
-   Node Version (`node -v`): 15.5.1
-   NPM Version (`npm -v`): 7.3.0
-   OS: MacOS 11.2.3

### Steps To Reproduce: 
I haven't tried creating a new project to see if it has the same problem.

**Comment by KonradDeskiewicz** [comment]
Same issue here without Sail... watch detects only the first change after starting, and then it looks that it recompiles something but no changes in output

**Comment by eveames** [comment]
Mine is working again, last I checked. I don't think I did something that fixed it. Would be nice to know what's going on...

**Comment by josermzte** [comment]
Same here!

**Comment by djzara** [comment]
I do believe I'm having the same issue. 



    Laravel Mix Version: 6.0.16
    Node Version (node -v): 12.16.0
    NPM Version (npm -v): 7.7.6
    OS: MacOS 11.2.3

On the side, and I cannot confirm if this is related, if I have an error during compilation via typescript mode, mix ends up in an endless loop of compile error compile error so on and so forth

**Comment by wayne-thorley** [comment]
Same here. No changes being reflected after the first compilation...
```
├── @babel/preset-react@7.13.13
├── axios@0.21.1
├── laravel-mix@6.0.13 <----------------
├── laravel-mix-merge-manifest@2.0.0
├── postcss@8.2.8
├── react@16.14.0
├── react-dom@16.14.0
├── react-router-dom@5.2.0
├── resolve-url-loader@3.1.2
└── tailwindcss@2.0.4
```

**Comment by thecrypticace** [comment]
If you're using Docker you're almost certainly gonna have to use watch-poll. File system events don't flow into Docker and that's not something we can fix. If watch-poll is not working that sounds like files aren't syncing in the Docker volume.

I'm not sure what (if anything) we can do about this. For now, I'd recommend running mix outside of Docker when developing.

**Comment by djzara** [comment]
> If you're using Docker you're almost certainly gonna have to use watch-poll. File system events don't flow into Docker and that's not something we can fix. If watch-poll is not working that sounds like files aren't syncing in the Docker volume.
> 
> I'm not sure what (if anything) we can do about this. For now, I'd recommend running mix outside of Docker when developing.

@eveames seems to be on docker, however I'm not and I don't think everyone else is either that have had this issue.

**Comment by thecrypticace** [comment]
@djzara Can you provide a repo that reproduces the problem you're seeing with watching files? (and the typescript one too? — that's very odd)

**Comment by djzara** [comment]
@thecrypticace yeah it's the weirdest thing I've seen. I've gone through several things to figure out what might cause it to happen. Regular ol' watch, compiles twice with only one compilation doing anything, and then that happens any time i save. No docker, not even a vagrant box piping in fs events. If there's an error you can end up in an infinite loop of compilations. 

As soon as I can I'm gonna setup a project from scratch and see if I can manage to reproduce it by gradually incorporating what I've done in to it and then get back to ya.

**Comment by wayne-thorley** [comment]
I'm not on docker & watch poll does the same thing...

**Comment by fridzema** [comment]
Same thing here, anybody has any updates on this?

**Comment by djzara** [comment]
@wayne-thorley  or @fridzema  have you been able to reliably reproduce this in a new project from scratch? I'm trying to figure out exactly what I'd done to make this happen, what thing I can say is that beyond mix I was using the vue composition api with typescript. Were you all? If that's the case I can see this being something to bring up to the vuejs or typescript folks, but even then I'm not sure how mix would even allow that kind of behavior.

**Comment by fridzema** [comment]
@djzara i can confirm this issue doesnt exist on a fresh install (at least for me). I am currently trying to reproduce the issue in a seperate repo https://github.com/fridzema/mix-watch-issue but eveything is working fine till now.

**Comment by djzara** [comment]
@fridzema thanks for that. I noticed this issue with typescript + vuex + composition API, so I'll try and hit it from that direction. Hopefully I can get a repo setup soon, crazy stuff in life. Versions for anyone interested:

- Laravel 8.35.0
- Laravel Mix 6.0.6
- Yarn 1.22.10
-  vue 3.0.7
- vuex ^4.0.0
- typescript 4.2.3
- tailwindcss 2.0.3
- postcss 8.1.14
- postcss-import 14.0.1

Figured I'd include some of the other items compiled by mix just in case. This is from the project with the problem. I'll keep these in mind when I can get my repro repo setup (crossing my fingers it's soon)

**Comment by thecrypticace** [comment]
I've been writing a rather significant amount of typescript lately and I've also noticed some oddities around watching files & recompilation. It's like webpack sees the changes and re-runs but ts-loader either doesn't compile again or doesn't write the files to disk. The thing here is it's not easily reproducible — it just happens seemingly at random sometimes. I'm not certain this is a Mix issue or if it's something we can work around. 🤔

**Comment by djzara** [comment]
I'm in about the same boat @thecrypticace . I don't think it's necessarily a mix issue since it's just wrapping webpack, but it's also in charge of the webpack configuration when you're running it in ts mode (without much in the way of any other modification of the webpack config) so it could definitely be a webpack issue, or even the version of webpack that mix is using? I wouldn't mind poking around their repo and issues as well to see what's up. Maybe a cross reference to help.

**Comment by wayne-thorley** [comment]
I couldn't reproduce myself on a fresh install using the same package versions.
@djzara - It's not related to typescript on my end as I don't use it.
Weirdly, just like @eveames, my original issue seems to have resolved itself today. I haven't done anything that would / could have affected it but `watch` is now recompiling correctly every time ¯\_(ツ)_/¯ It's a weird one!

**Comment by iDoes123** [comment]
I had the same problem and tried absolutely everything. Eventually I found that the issue was with how I imported the Vue routes.

```
{
    path: '/test',
    name: 'test',
    component: () => import("../views/pages/test.vue")
},
```
I forgot to use a capital letter....

`component: () => import("../views/pages/test.vue")`

Should be:

`component: () => import("../views/pages/Test.vue")`

Apparantly when I run 'npm run watch' laravel mix is fine with importing my Test.vue eventhough I wrote 'test.vue'. Thats why I did see changes when I restarted the watch. However, after the initial start laravel mix uses dynamic loading and at that point it will ignore the changes if I do not use the capital letter. 

Watch will also ignore changes in components, after first run, if you import it without the correct use of capital or lowercase letters.

`import Button from '@/components/Button'`

I thought I'd post my mistake. Maybe some of you have the same issue?

**Comment by ntbrittain** [comment]
@iDoes123 I was running into this issue, and you nailed the cause, I had changed some file names and had the incorrect casing in an import (unrelated to vue). Fixing the casing solved the issue for me. Thanks!

**Comment by peter-brennan** [comment]
i'm running into the same issue but on a fresh install using `laravel new test-app`

Laravel v8.83.2 (PHP v7.3.24)
Mac OSX 12.2
Mamp Pro

package.json
`{
    "private": true,
    "scripts": {
        "dev": "npm run development",
        "development": "mix",
        "watch": "mix watch",
        "watch-poll": "mix watch -- --watch-options-poll=1000",
        "hot": "mix watch --hot",
        "prod": "npm run production",
        "production": "mix --production"
    },
    "devDependencies": {
        "axios": "^0.21",
        "laravel-mix": "^6.0.6",
        "lodash": "^4.17.19",
        "postcss": "^8.1.14"
    }
}
`

**Comment by skyyip** [comment]
`npm run watch-poll` saved my life,
Check the detail here: https://laravel.com/docs/5.4/mix#running-mix

**Comment by trevormiles** [comment]
What fixed it for me was similar to what @ghost mentioned.

I had an `Index.vue` file inside of a folder called `EnrollForm`, which I was importing like so:

`import EnrollForm from '../../EnrollForm'`

When I started up `npm run watch`, my changes were correctly reflected in the browser only for two compiles, and then the compiler would say "Build successful" but my changes were not reflected in the browser.

Per @ghost's comment, I tried changing the import statement to this:

`import EnrollForm from '../../EnrollForm/Index.vue'`

And then my changes began to be reflected in the browser every time the compiler ran 🤷‍♂️ 

Thanks @ghost!

**Comment by yazdanba98** [comment]
i just looking for some reason for my npm run watch why not apply assets every changes and its working like npm run dev if you have same issue as mine just i have to say some info 
1. i use laravel 8 
2. i use laravel-html version of vuexy

so for that problem i check my package.json and as i know that before my package.json missing dev script so i find it in my other projects and add it at first item of scripts ( in package.json) 

  "scripts": {
      "dev": "npm run development",
      .
      .
      .

now its working just normal and apply all assets after every change 


and again i'm not sure about what's purpose in this issue
